### PR TITLE
lib: Get rid of buggy filesystem wrapping

### LINKED ISSUE #7974: [Android] Renaming file uppercase causes both files to be deleted
**What happened**
I renamed test.txt to TEST.txt and both files were deleted.

**What operating system, operating system version and version of Syncthing you are running**
jpc device:
v1.18.2, Linux (64-bit Intel/AMD)

**The same for other connected devices, where relevant**
jphone device:
Syncthing Version: v1.18.2
Syncthing-Android Version: 1.18.2

**Screenshot if the issue concerns something visible in the GUI**
syncth

### PR DESCRIPTION
**Pull request by imsodin**
Investigating https://github.com/syncthing/syncthing/issues/7974 I found that the log-filesystem is incorrectly wrapped if `walkfs` debug facility is used. And fixing that I further noticed case-fs is also badly wrapped: It's outside of `walkFilesystem`, meaning walking doesn't use the case filesystem. Also there's code in `mtimeFS.Walk` to make walk aware of the mtime-ness - I removed that and instead now `walkFilesystem` wraps `mtimeFS`.  All this wrapping after a filesystem is created is ugly and error prone. Thus I removed the separate constructors for case and mtime fs, and instead made those options to `NewFilesystem`. Like that the "correct layering" happens right there in a single constructor, no re-wrapping later on.

So basically the largest part of the diff is just the refactor for mtime/case-fs creation. The functional changes are:

 - Move `caseFilesystem` from outside to inside `walkFilesystem`.
 - Move `mtimeFS` from outside to inside `walkFilesystem`, removing the need for `mtimeFS.Walk`.
 - As a side-effect of the first item: Fs logging now shows the correct caller, instead of the `caseFilesystem` wrapper method.

I have yet to test if this fixes #7974 (I doubt it).

### CODE REVIEW

File: lib/fs/filesystem.go [hunk 1]
```diff
@@ -202,10 +202,28 @@ var IsPathSeparator = os.IsPathSeparator
...
+		switch opt.(type) {
+		case *OptionDetectCaseConflicts:
+			caseOpt = opt
+		case *optionMtime:
+			mtimeOpt = opt
+		default:
+			opts[i] = opt
+			i++
+		}
+	}
```
Comment (calmh) [inline lib/fs/filesystem.go:225]: Maybe something here to `opts = opts[:i]` or we might have some dangling extra options?

File: lib/config/folderconfiguration.go [hunk 2]
```diff
@@ -43,17 +44,23 @@ func (f FolderConfiguration) Copy() FolderConfiguration {
 }
 
 func (f FolderConfiguration) Filesystem() fs.Filesystem {
```
Comment (calmh) [inline lib/config/folderconfiguration.go:46]: Is this one really used? It seems possibly error prone to keep around, maybe if it's only used from tests and stuff we could make this the main thing and have callers do `Filesystem(nil)` when they specifically want a non-mtime version of it?

**Review by calmh — COMMENTED** [review]
Good

File: lib/config/folderconfiguration.go (hunk 2 above)
Comment (imsodin) [reply lib/config/folderconfiguration.go]: It is used in quite a few places. Mostly for "simple operations" where mtime-handling doesn't matter (.stfolder, getting the filesystem uri, ...). Just the file versioner uses mod-time, but is not an mtime-filesystem. That's affected by the same issues around mtime as the "main folder filesystem", so likely it would be good to have it there too - haven't thought it trough.

File: lib/fs/filesystem.go [hunk 3]
```diff
@@ -221,6 +239,17 @@ func NewFilesystem(fsType FilesystemType, uri string, opts ...Option) Filesystem
 		}
 	}
 
+	// Case handling is the innermost, as any filesystem calls by wrappers should be case-resolved
+	if caseOpt != nil {
+		fs = caseOpt.apply(fs)
+	}
+
+	// mtime handling should happen inside walking, as filesystem calls while
+	// walking should be mtime-resoved too
```
Comment (rasa) [inline lib/fs/filesystem.go:249]:
```suggestion
	// walking should be mtime-resolved too
```

**Review by calmh — APPROVED** [review]

**Comment by es20490446e** [comment]
Thanks for the fix 👍

**Comment by imsodin** [comment]
@es20490446e Ufortunately this is not a fix for https://github.com/syncthing/syncthing/issues/7974, at least not until proven otherwise. It's just something I noticed and needed to fix for investigating that issue - getting to the bottom of it is still outstanding.
